### PR TITLE
Add verification of dependency metadata

### PR DESCRIPTION
**Pull request by melix**
### Context

Add dependency verification of metadata

### CODE REVIEW

**Comment by melix** [comment]
Build is failing on a performance regression which is expected. Will temporarily disable the test before merging then rebaseline. A subsequent PR will try to improve performance.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ModuleDescriptorHashCodec.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        return true;
+    }
+
+    @Override
+    public int hashCode() {
+        return super.hashCode();
```
Comment (jjohannes) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ModuleDescriptorHashCodec.java]: This is probably ok. I was just wondering: as all instances of the class are "equal", should this be a constant hash code?

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/metadata/AbstractRepositoryMetadataSource.java [hunk 2]
```diff
@@ -79,12 +85,26 @@ private S parseMetaDataFromArtifact(String repositoryName, ComponentResolvers co
...
+                sources.add(new DefaultMetadataFileSource(artifact.getId(), metadataArtifactFile, findSha1(metaData, metadataArtifactFile)));
+                context.appendSources(sources);
+                return metadata;
             }
         }
         return null;
     }
 
+    static BigInteger findSha1(ExternalResourceMetaData metaData, File artifact) {
+        HashValue sha1 = metaData.getSha1();
```
Comment (jjohannes) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/metadata/AbstractRepositoryMetadataSource.java]: Out of curiosity, and maybe I just missed it: What happens if the published sha is wrong? What kind of error will the use see at which moment?

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ModuleDescriptorHashCodec.java (hunk 1 above)
Comment (melix) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ModuleDescriptorHashCodec.java]: This code is a bit messy, not implementing `equals` makes composite build fail, but you don't _have to_ implement `hashCode`. I'm not sure why this is that way, surely we can find better ways.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/metadata/AbstractRepositoryMetadataSource.java (hunk 2 above)
Comment (melix) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/metadata/AbstractRepositoryMetadataSource.java]: Here I think it just means the file would be stored in the "wrong" directory.

**Pull request merged**